src/dos/cdrom_image.cpp

```diff
@@ -182,6 +188,57 @@ bool CDROM_Interface_Image::AudioFile::seek(Bit32u offset)
 	return result;
 }
 
+bool CDROM_Interface_Image::AudioFile::read(uint8_t *buffer,
+                                            const uint32_t offset,
+                                            const uint32_t requested_bytes)
+{
+	// Check for logic bugs
+	assertm(buffer != nullptr, "buffer needs to be allocated but is the nullptr [Bug]");
+	assertm(sample != nullptr, "Audio sample needs to be valid, but is the nullptr [Bug]");
+
+	// Guard again valid but no-op case
+	if (requested_bytes == 0)
+		return true;
+
+	// We support DAE from 16-bit, stereo, 44 kHz tracks. If the track doesn't conform to
+	// this, then inform the user. Also, we allow up to 10 DAE-attempts before informing
+	// the user - because some CD Player software will query this interface before using
+	// CDROM-directed playback (not DAE), therefore we don't want to fail in those cases.
+	if (getRate() != REDBOOK_PCM_FRAMES_PER_SECOND || getChannels() != REDBOOK_CHANNELS) {
+		static uint8_t dae_attempts = 0;
+		if (dae_attempts++ > 10) {
+			E_Exit("\n"
+			       "CDROM: Digital Audio Extration (DAE) was attempted with a %s %u kHz\n"
+			       "       track, but DAE is currently only compatible with stereo %u kHz\n"
+				   "       tracks.",
+			       getChannels() == 2 ? "stereo" : "mono",
+			       getRate(),
+			       REDBOOK_PCM_FRAMES_PER_SECOND);
+		}
+		return false; // we always correctly return false to the application in this case. 
+	}
+
+	// Seek, but only if we have to
+	if (position != (offset - requested_bytes))
+		if (!seek(offset))
+			return false;
+	position = offset;
+
+	uint32_t decoded_frames = 0;
+	const uint32_t requested_frames = requested_bytes / BYTES_PER_REDBOOK_PCM_FRAME;
+	while (decoded_frames < requested_frames
+	       && !(sample->flags & (SOUND_SAMPLEFLAG_ERROR | SOUND_SAMPLEFLAG_EOF))) {
```

Comment:
These lines were the only ones, that gave me pause when reading new code.

How about rewriting it this way:

	while (decoded_frames < requested_frames) {
		if (sample->flags & (SOUND_SAMPLEFLAG_ERROR | SOUND_SAMPLEFLAG_EOF))
			break;

It should compile to the same code and seems a bit easier to understand.

(This is only a suggestion, feel free to ignore it if you prefer to keep while loop as it is now)

Reply:
Yes, that is an readability improvement - thanks @dreamer 
Will drop it in, in my morning.

Reply:
Done